### PR TITLE
feat: add topological sort, support for previewing changes

### LINKED ISSUE #199: Parallelize per package tasks
**Is your feature request related to a problem? Please describe.**

There are a few tasks which run per package. Currently we use Promise.all which doesn't actually perform parallel processing.

**Describe the solution you'd like**

Abstract Promise.all usage for parallelization helper, and explore worker threads.

### LINKED ISSUE #257: Run lifecycle hooks in topological sort order via CLI flag
**Is your feature request related to a problem? Please describe.**

If package A depends on package B, there should be a guarantee that package B is built before package A's prepack is run.

**Describe the solution you'd like**

A new CLI flag to enable topological sort.

**Describe alternatives you've considered**

**Additional context**

Both lerna and yarn workspaces supports topological sort. Yarn has 2 options for this sort, one based on dependencies, and one based on dependencies+devDependencies. We can also support both.

### PR DESCRIPTION
**Pull request by noahnu**
## Description

Few changes in this PR:
- Adds `--topological`, `--topological-dev`, `--jobs` cli flags similar to yarn workspace tools.
- Removes dependency on example-monorepo in tests, and adds a helper to generate test monorepos.
- Adds `--force-write-change-files` cli flag to write changelog and changeset even in dry run mode. Useful for previewing changes.
- Bug fixes to get dependents algorithm.
- Bug fixes to changelog template context.
- Updates to yarn 2.4.1.

## Related Issues

- Closes #257, adds topological sort option
- Closes #199, parallelize per package tasks

## Checklist

- [x] I agree to abide by the [Code of Conduct](https://github.com/tophat/monodeploy/blob/master/CODE_OF_CONDUCT.md).
- [x] This PR has sufficient documentation.
- [x] This PR has sufficient test coverage.
- [x] This PR title satisfies semantic [convention](https://www.conventionalcommits.org/en/v1.0.0/#summary).

## Additional Comments

Patched jest to fix a bug with resolving.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -46,6 +46,7 @@
         "@babel/preset-env": "^7.13.9",
         "@babel/preset-typescript": "^7.13.0",
         "@commitlint/cli": "^11.0.0",
+        "@monodeploy/test-utils": "link:./testUtils",
```
Comment (noahnu) [inline package.json:49]: I reserved the `@monodeploy` scope in npm. We could change our package names if we wanted to.

Comment (mcataford) [reply package.json:49]: Oh neat! Do you know if there's a standard procedure when renaming a package like that? I assume that if we were to move, we'd publish a nochange version of `monodeploy-*` packages to NPM that mention linking to the new one?

File: packages/cli/src/cli.ts [hunk 2]
```diff
@@ -93,20 +98,21 @@ if (argv.logLevel !== undefined && argv.logLevel !== null) {
...
-            push: argv.push ?? undefined,
+            push: argv.push,
         },
         conventionalChangelogConfig:
             argv.conventionalChangelogConfig ?? undefined,
         changesetFilename: argv.changesetFilename ?? undefined,
         changelogFilename: argv.prependChangelog ?? undefined,
         access: argv.access ?? undefined,
-        persistVersions: argv.persistVersions ?? undefined,
+        persistVersions: argv.persistVersions,
```
Comment (mcataford) [inline packages/cli/src/cli.ts:108]: **[question]**

Curious as to why we don't null-coalesce `persistVersion` and `push` to be `undefined` if not provided in cli args, unless it's a leftover of testing?

File: testUtils/setupMonorepo.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+    await context.project.install({
+        cache: await Cache.find(context.configuration),
+        report: new ThrowReport(),
+    })
+
+    return context
+}
+
+export async function withMonorepoContext(
```
Comment (mcataford) [inline testUtils/setupMonorepo.ts:95]: This is rad. :rocket:

**Review by mcataford — COMMENTED** [review]
Your toposort algo makes sense to me, but I'll do some pen-and-papering later today to fully wrap my head around it (+ the groups). As to `--topological-dev` v `--topological`, I would this interesting tidbit on Discord:

> I'd say no - --topological-dev is special because there can be cases when strict topological sorting isn't possible - hence why we have --topological, which is more lax

I also assume that in much larger monorepos, if you want a quick traversal and know that not including devDependencies in it won't cause problems, then ignoring the dev packages is a good path forward. :thinking:

File: package.json (hunk 1 above)
Comment (noahnu) [reply package.json:49]: if we keep the CLI as monodeploy, we wouldn't really have to change anything else. No one is using the supporting packages. So the CLI would be "monodeploy" and the other packages would be `@monodeploy/types` etc..

File: packages/cli/src/cli.ts (hunk 2 above)
Comment (noahnu) [reply packages/cli/src/cli.ts:123]: persistVersions has a default value so it's never null

File: packages/node/src/monodeploy.test.ts [hunk 4]
```diff
@@ -576,45 +586,55 @@ describe('Monodeploy Lifecycle Scripts', () => {
...
-            await cleanup('pkg-4')
+        // pkg-5 depends on pkg-4, so it'll be bumped as a dependant
+        // pkg-3 is bumped because it depends on pkg-2
+        // pkg-6 is bumped because it depends on pkg-3
+        expect(mockGit._getPushedTags_()).toEqual([
+            'pkg-2@0.2.0',
+            'pkg-3@0.0.2',
+            'pkg-4@0.1.0',
+            'pkg-5@0.0.2',
+            'pkg-6@0.0.2',
```
Comment (noahnu) [inline packages/node/src/monodeploy.test.ts:623]: This is failing. For some reason pkg-6 is not being updated. Shouldn't it be updated as a dependent of a dependent?

**Pull request merged**